examples/rust/opensk/src/user_presence.rs

```diff
@@ -12,22 +12,75 @@
 // See the License for the specific language governing permissions and
 // limitations under the License.
 
-use opensk_lib::api::user_presence::UserPresence;
+use alloc::rc::Rc;
+use core::cell::Cell;
+use core::time::Duration;
+
+use opensk_lib::api::user_presence::UserPresenceError::{Fail, Timeout};
+use opensk_lib::api::user_presence::{UserPresence, UserPresenceResult};
+use wasefire::button::{self, Listener};
+use wasefire::clock::{Oneshot, Periodic};
+use wasefire::{clock, led, scheduling};
 
 use crate::WasefireEnv;
 
 impl UserPresence for WasefireEnv {
-    fn check_init(&mut self) {
-        todo!()
-    }
+    fn check_init(&mut self) {}
 
-    fn wait_with_timeout(
-        &mut self, timeout_ms: usize,
-    ) -> opensk_lib::api::user_presence::UserPresenceResult {
-        todo!()
-    }
+    fn wait_with_timeout(&mut self, timeout_ms: usize) -> UserPresenceResult {
+        // Make sure there is at least one button and one LED.
+        let num_buttons = button::count();
+        let num_leds = led::count();
+        if num_buttons == 0 || num_leds == 0 {
+            return Err(Fail);
+        }
+
+        // Start listening for button presses.
+        // TODO: Should we be checking for any button pressed or only index 0?
+        let button_pressed = Rc::new(Cell::new(false));
+        for index in 0 .. num_buttons {
+            let button_pressed = button_pressed.clone();
+            let handler = move |state| match state {
+                button::Pressed => button_pressed.set(true),
+                button::Released => (),
+            };
+            let listener = Listener::new(index, handler);
+            listener.leak();
+        }
+
+        // Start a periodic timer blinking the LED.
+        // TODO: Similarly, should we be blinking all LEDs or only index 0?
+        let blinking = Rc::new(Cell::new(false));
```

Comment:
We don't need this variable (and you probably meant to initialize it to true). It is enough to do:

```rust
let blink_timer = clock::Timer::new(|| led::set(0, !led::get(0)));
led::set(0, led::On);
blink_timer.start(Periodic, Duration::from_millis(200));
// [...]
scheduling::wait_until(|| button_pressed.get() || timeout.get());
drop(blink_timer);
led::set(0, led::Off);
```

Reply:
Done.